csrc/fusion.cpp

```diff
@@ -824,9 +828,9 @@ void Fusion::aliasOutputToInput(
       .aliased_io = input,
       .hide_output = !output->isFusionOutput()};
```

Comment:
Can we enforce aliasOutputToInput to happen before addOutput (if addOutput is used at all for unhiding the output)? This would simplify the logic here to something like
```suggestion
      .hide_output = true};
  addOutputHelper(output);
```
and possibly other functions.

Reply:
I'm not a big fan on adding restriction on python API. i.e. less restriction == less customer facing doc and more freedom on integration. 

But we can throw an informative error message on that, so I guess this is still a fair conversation. But how strong are you on this? I'm trying to push it back unless you have a firm stance :stuck_out_tongue_winking_eye:

Reply:
> and possibly other functions.

My argument is that, the suggested change won't actually simplify the other function by much. It's actually the subsequent `addOutput` which requires updating the `hide_output` field and removing the existing entry in outputs_.

Reply:
SGTM as is. 

Overall, I'm unhappy with aliasOutputToInput and addOutput having overlapping responsibilities -- both may call addOutputHelper and may modify io_alias_. It's not immediately clear to the user of nvFuser how many times `out` is added in the following cases. 

```
aliasOutputToInput(out, ...);
```
=> once

```
aliasOutputToInput(out, ...);
addOutput(out);
```
=> once

```
addOutput(out);
aliasOutputToInput(out, ...);
```
=> once

```
addOutput(out);
aliasOutputToInput(out, ...);
addOutput(out);
```
=> twice

and things become even more complicated with removeOutput. 

I don't have an easy solution. Just for the record, we may consider to add a finalization phase (if we don't have already?) when building a fusion. Then, change code to do the following
1. aliasOutputToInput only marks `io_alias_` without changing the output list or modifying hide_output
2. addOutput only adds output to list without modifying io_alias_
3. During finalization: 
   ```
   for each key in io_alias_:
     if key is an output: 
       io_alias_[key].hide = false
     else:
       io_alias_[key].hide = true
       addOutput
   // This way, each key in io_alias_ are guaranteed to be an output for implementation purpose. 
   ```

This way, aliasOutputToInput and addOutput have non-overlapping responsibilities. The dirty implementation detail -- an alias has to be an output -- is handled by finalization, just once and at the end.

Reply:
I think this was resolved during our offline conversation.

I agree that we could use some re-naming here to avoid confusion.

`addOutput(out);` should be the only API that defines a returned value in fusion.
`aliasOutputToInput(out, ...);` should be renamed to something appropriately now since it doesn't define a returned value, though the implementation requires something added to `outputs_`.

`finalization phase` is a reasonable alternative in implementation. Although the logic in this PR is sufficient to resolve that when the APIs are called, and we do not have to defer that at the end.